### PR TITLE
Storybook: Opt in to story store v7

### PR DESCRIPTION
**Pull request by stokesman**
—https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#story-store-v7

## How?
Making the required configuration changes. I'm not sure if I'm missing anything else though.

## Testing Instructions
1. Run storybook and make sure all stories are available

### CODE REVIEW

File: storybook/main.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 const stories = [
 	process.env.NODE_ENV !== 'test' && './stories/**/*.@(js|tsx|mdx)',
 	'../packages/block-editor/src/**/stories/*.@(js|tsx|mdx)',
-	'../packages/components/src/**/stories/*.@(js|tsx|mdx)',
+	'../packages/components/src/**/stories/index.@(js|tsx|mdx)',
```
Comment (stokesman) [inline storybook/main.js]: Without this change the build seems to work fine but produces warnings about various files not having a default export. This change to the more specific pattern seemed the laziest way to avoid those warnings and I don't think it excludes anything. I'm open to any suggestions about doing this a different way. The two folders with files that produce warnings are:
- https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/navigation/stories
- https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/popover/stories

Comment (mirka) [reply storybook/main.js]: And also [ToolsPanel](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/tools-panel/stories).

It kind of seems that ideally this pattern should be handled in Storybook itself (cf. https://github.com/storybookjs/storybook/issues/16598). But otherwise, it's probably either the `index` way, or moving sub-stories into a subfolder way. Someone may be confused when a story file doesn't get picked up as expected, but I don't see that happening very often. I think we can go with `index` for now and revisit if it causes issues.

**Review by mirka — COMMENTED** [review]
Wow, the load time improvement is very perceivable! Thanks for taking the initiative on this one. It was on my mind as well, after seeing #42377 😆 

It looks like we're going to need a bit more testing of intentional/unintentional globals, so we don't get unexpected behavior. I noticed that a global `font-family` is not being loaded anymore (e.g. see AnglePickerControl), which is likely due to #42414 not being loaded on initial load.

My inclination at this point is for `font-family` to be set as a global rule, instead of in each component. However, `box-sizing` resets will probably need to be added on a per-component basis (#42415).

For this PR though, I guess an intermediary step could be to move the global `font-family`, `font-size`, and `box-sizing` reset to a proper global style instead of the [accidental leaky file](https://github.com/WordPress/gutenberg/issues/42414).

File: storybook/main.js (hunk 1 above)
Comment (ciampo) [reply storybook/main.js]: > I think we can go with index for now and revisit if it causes issues.

Sounds like a good plan

**Comment by ciampo** [comment]
Thank you for working on this, @stokesman !



> My inclination at this point is for `font-family` to be set as a global rule, instead of in each component. 

Since I believe the intention is for all components to use the same `font-family`, we should probably aim at adding explicitly the `font-family: inherit` rule on all components? Alternatively, we could also use a CSS variable.

> However, box-sizing resets will probably need to be added on a per-component basis (https://github.com/WordPress/gutenberg/issues/42415).

Regarding `box-sizing`, I'd like to explore the possibility of setting all components to `border-box` (as discussed in https://github.com/WordPress/gutenberg/issues/42415#issuecomment-1184830596)

> It looks like we're going to need a bit more testing of intentional/unintentional globals, so we don't get unexpected behavior.
> ...
> For this PR though, I guess an intermediary step could be to move the global font-family, font-size, and box-sizing reset to a proper global style instead of the https://github.com/WordPress/gutenberg/issues/42414.

Agreed. We should look at the leaky styles and see which ones would potentially need to be replicated temporarily in this PR — `font-family`, `font-size`, and `box-sizing` are definitely on the list.

**Comment by mirka** [comment]
> Since I believe the intention is for all components to use the same `font-family`, we should probably aim at adding explicitly the `font-family: inherit` rule on all components? Alternatively, we could also use a CSS variable.

My concern is that this would add unnecessary lines of code and CSS specificity, since `font-family` is already an inherited property by default. The only places where we "need" to set it explicitly are the elements that have user agent styles (#37555).

**Comment by ciampo** [comment]
> My concern is that this would add unnecessary lines of code and CSS specificity, since `font-family` is already an inherited property by default. The only places where we "need" to set it explicitly are the elements that have user agent styles (#37555).

Makes sense

**Comment by mirka** [comment]
The global CSS issues will be handled in #42747.

File: storybook/main.js (hunk 1 above)
Comment (ciampo) [reply storybook/main.js]: [`NavigableContainer` too](https://github.com/WordPress/gutenberg/tree/2afc39560a8e28f1156ceb045ebc3d50143e8d20/packages/components/src/navigable-container/stories)

**Comment by mirka** [comment]
@stokesman I would really like this PR to land 😊 I don't think there are any blockers anymore, and it seems to be working ok after a rebase. Are you available to land it, or would you rather I take over?

File: storybook/main.js (hunk 1 above)
Comment (stokesman) [reply storybook/main.js]: I checked for any directories named "stories" that are without index files and found five:
```
./packages/components/src/ui/context/stories,
./packages/components/src/date-time/stories,
./packages/components/src/duotone-picker/stories,
./packages/components/src/navigable-container/stories,
./packages/components/src/utils/hooks/stories'
```
So I think we'd need to add an index file or else they'd be overlooked by this change.

**Comment by stokesman** [comment]
> I would really like this PR to land

Me too! And I'd really like if you'd like to take over. I left https://github.com/WordPress/gutenberg/pull/42486#discussion_r1019535261 about what I think is the only remaining todo here. Thanks for checking in and for the great work you did that removed the blockers for this!

**Comment by mirka** [comment]
Gotcha, I'll take over then. Thank you!

File: storybook/main.js (hunk 1 above)
Comment (mirka) [reply storybook/main.js]: The "no default export" errors are not appearing anymore for some reason 🤔 So I reverted the glob change for now.

If it reappears again, I think we could move to an approach where the glob is just for the root-level files in `stories/`, and then move the non-story files into a `stories/utils/` subfolder or something.

Comment (mirka) [reply storybook/main.js]: My bad, I forgot those errors were emitted at build-time, not at runtime (I was checking in devtools console) 🙈 I did end up moving the files to `utils/` subfolders.

File: packages/components/tsconfig.json [hunk 2]
```diff
@@ -37,8 +37,8 @@
 		"src/**/*.ios.js",
 		"src/**/*.native.js",
 		"src/**/react-native-*",
-		"src/**/stories/**.js", // only exclude js files, tsx files should be checked
-		"src/**/test/**.js", // only exclude js files, ts{x} files should be checked
+		"src/**/stories/**/*.js", // only exclude js files, tsx files should be checked
+		"src/**/test/**/*.js", // only exclude js files, ts{x} files should be checked
```
Comment (mirka) [inline packages/components/tsconfig.json:41]: Wrong globbing!

**Comment by mirka** [comment]
Ok, should be good now. If anyone could do a quick smoke test and approve this PR, we can merge.

tl;dr for those watching: Basically the load time should be perceivably faster compared to before. (That means [LCP](https://web.dev/lcp/), not the build time unfortunately 😅)

**Review by ciampo — APPROVED** [review]
Works well! (and loads fast!) 🚀 

Thank you @stokesman and @mirka for working on this 🙏
